sympy/matrices/determinant.py

```diff
@@ -0,0 +1,979 @@
+    [ 4, -2],
+    [-3,  1]])
+    >>> adjugate(M)
+    Matrix([
+    [ 4, -2],
+    [-3,  1]])
+
+    See Also
+    ========
+
+    cofactor_matrix
+    sympy.matrices.common.MatrixCommon.transpose
+    """
+
+    return M.cofactor_matrix(method=method, dotprodsimp=dotprodsimp).transpose()
+
+
+# This functions is a candidate for caching if it gets implemented for matrices.
+def _charpoly(M, x='lambda', simplify=_simplify, dotprodsimp=None):
+    """Computes characteristic polynomial det(x*I - M) where I is
+    the identity matrix.
+
+    A PurePoly is returned, so using different variables for ``x`` does
+    not affect the comparison or the polynomials:
+
+    Parameters
+    ==========
+
+    x : string, optional
+        Name for the "lambda" variable, defaults to "lambda".
+
+    simplify : function, optional
+        Simplification function to use on the characteristic polynomial
+        calculated. Defaults to ``simplify``, if ``dotprodsimp`` is ``True``
+        then this is ignored.
+
+    dotprodsimp : bool, optional
+        Specifies whether intermediate term algebraic simplification is used
+        to control expression blowup during matrix multiplication. If this
+        is true then the simplify function is not used.
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> from sympy.matrices import charpoly
+    >>> from sympy.abc import x, y
+    >>> M = Matrix([[1, 3], [2, 0]])
+    >>> M.charpoly()
+    PurePoly(lambda**2 - lambda - 6, lambda, domain='ZZ')
+    >>> charpoly(M)
+    PurePoly(lambda**2 - lambda - 6, lambda, domain='ZZ')
+    >>> M.charpoly(x) == M.charpoly(y)
+    True
+    >>> M.charpoly(x) == M.charpoly(y)
+    True
+
+    Specifying ``x`` is optional; a symbol named ``lambda`` is used by
+    default (which looks good when pretty-printed in unicode):
+
+    >>> M.charpoly().as_expr()
+    lambda**2 - lambda - 6
+
+    And if ``x`` clashes with an existing symbol, underscores will
+    be prepended to the name to make it unique:
+
+    >>> M = Matrix([[1, 2], [x, 0]])
+    >>> M.charpoly(x).as_expr()
+    _x**2 - _x - 2*x
+
+    Whether you pass a symbol or not, the generator can be obtained
+    with the gen attribute since it may not be the same as the symbol
+    that was passed:
+
+    >>> M.charpoly(x).gen
+    _x
+    >>> M.charpoly(x).gen == x
+    False
+
+    Notes
+    =====
+
+    The Samuelson-Berkowitz algorithm is used to compute
+    the characteristic polynomial efficiently and without any
+    division operations.  Thus the characteristic polynomial over any
+    commutative ring without zero divisors can be computed.
+
+    See Also
+    ========
+
+    det
+    """
+
+    if not M.is_square:
+        raise NonSquareMatrixError()
+
+    if dotprodsimp:
+        simplify = lambda e: e
+
+    berk_vector = _berkowitz_vector(M, dotprodsimp=dotprodsimp)
+    x = _uniquely_named_symbol(x, berk_vector)
+
+    return PurePoly([simplify(a) for a in berk_vector], x)

```

Comment:
Not sure about this but is it possible to move this simplification to the Matrix method so we don't need that simplify argument here? Then maybe the signature here could just be simplify=True/False with the effect that True unables dotprodsimp. The docstring should be vague about exactly what that does.

Reply:
It is possible and of course desirable to do lots of tweaks to make simplification args less messy but I was specifically trying to not change the functions or signatures in this PR, just relocate them. And if you say "Matrix method" do you mean you want the `Matrix.charpoly()` to possibly work different from `charpoly(Matrix)` and change the sig?

Reply:
I would say don't get greedy and mix goals, first lets isolate the funcs then improve once that is settled. I refer again to #18253 me responding to @sylee:

> EDIT: Thinking about this it is an interesting idea, though how to standardize it?

> > Organize options like method='bareiss_simplified', simpoptions={'after_op_1': foo, 'after_op_2': bar} such that it is maximally-configurable, and try to make dotprodsimp or such fit into this. (And we should clearly document something like what opeartion 1 does in the intermediate steps, and such)

> EDIT2: On the previous edit, the simpoptions dict could contain iszerofunc, dotprodfunc, simpfunc and maybe postsimpfunc. We could standardize some prepared dicts for quick use such as Matrix.nosimp, Matrix.dotprodsimp, Matrix.fullsimp and maybe some others which could be specified by either a dict object or a string ('none', 'dotprod', 'full') for the simpoptions or just simpopts keyword. Any function which does not use one of those simplification steps or functions would just ignore it so instead of a call like:
```
LUdecomposition_Simple(iszerofunc=someiszerofunc, simpfunc=somesimpfunc, dotprodsimp=somedotprodsimp)
```
it would be:
```
LUdecomposition_Simple(simpopts='full')
```
This is one idea for collapsing all those separate simplification options into one argument. But whatever the final form I think it should be well thought out and comprehensive, not just a small tweak or hack here and there.

Reply:
I have probably the idea to be 'minimal' multiple times before.

A good point of introducing a new API is that we can begin from the scratch and don't necessarily have to keep all the stuff that were badly designed or too complicated to explain for users.

I have only stated two options, either to be maximally controllable or either to be minimal, because I wasn't clear about you were proposing a lot of controls, or a simpler control in the previous discussions. 

But if you finally acknowledge the advantage of having minimal keywords like `simpopts='full'`, I would vote for that.

Reply:
I think minimal is good. The reason I ask about this particular example though is that it looks like it's just simplifying at the end. It's better to let the user do `trigsimp(M.charpoly())` than provide a redundant argument so they can do `M.charpoly(simplify=trigsimp)`.

Reply:
I agree with everything said, just resisting doing operational or sig changes in this particular PR to keep it as a relocation only.

Reply:
> I have only stated two options, either to be maximally controllable or either to be minimal, because I wasn't clear about you were proposing a lot of controls, or a simpler control in the previous discussions.

There is a possibility to be both, the `simpopts` arg could take string or dict, internally a string would be used to look up a pre-defined dict and that dict is what would actually be used. So if you want simplicity you do `M.charpoly(... simpopts='full' ...)` and for control you could do `M.charpoly(... simpopts={'iszerofunc': somefunc, `simpfunc`: somefunc, ...} ...)`.

Reply:
Here is a quick example of how I imagine the above would work:
```
from sympy.simplify.simplify import simplify as _simplify, dotprodsimp as _dotprodsimp

_noplamb          = lambda x: x

_simpopts_default = {}
_simpopts_none    = {'simplify': False, 'simpfunc': _noplamb, 'dotprodsimp': False}
_simpopts_full    = {'simplify': True, 'simpfunc': _simplify, 'dotprodsimp': True}
_simpopts_dotprod = {'dotprodsimp': True}

_simpopts_dicts   = {
  None:      _simpopts_default,
  False:     _simpopts_none,
  True:      _simpopts_full,
  'default': _simpopts_default,
  'none':    _simpopts_none,
  'full':    _simpopts_full,
  'dotprod': _simpopts_dotprod,
}

def _get_simpopts(simpopts, *args, **kwargs):
  simplify_as_func    = kwargs.get('simplify_as_func', False)
  dotprodsimp_as_func = kwargs.get('dotprodsimp_as_func', False)

  simpopts = _simpopts_dicts.get(simpopts, simpopts) # if 'simpopts' is dict it is used
  ret      = []

  for opt, default in zip(args[::2], args[1::2]):
    val = kwargs.get(opt, ret) # ret used as sentinel

    if val is ret:
      val = simpopts.get(opt, default)

    if opt == 'simplify':
      if simplify_as_func and not callable(val):
        val = _simplify if val else _noplamb

    elif opt == 'dotprodsimp':
      if dotprodsimp_as_func and not callable(val):
        val = _dotprodsimp if val else _noplamb

    ret.append(val)

  return ret
```
Now this:
```
def _det_LU(M, iszerofunc=_iszero, simpfunc=None, dotprodsimp=None):
```
would become:
```
def _det_LU(M, simpotps=None, **kwargs):
  iszerofunc, simpfunc, dotprodsimp = _get_simpopts(simpopts,
                                                    'iszerofunc', _iszero,
                                                    'simpfunc', None,
                                                    'dotprodsimp', None)
```
This would handle the difference between functions which use `simplify` as a bool and others which expect a function. The `kwargs` are there for backward compatibility.